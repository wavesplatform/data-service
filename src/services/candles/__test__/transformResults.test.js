const { groupBy, map, pipe, toPairs } = require('ramda');
const { addMissingCandles, transformCandle } = require('../transformResults');
const { candleMonoid } = require('../candleMonoid');
const { interval, Unit, CandleInterval } = require('../../../types');
const { floor, trunc } = require('../../../utils/date');
const { concatAll } = require('../../../utils/fp/concatAll');

const truncToMinutes = trunc(Unit.Minute);

const oneDayCandles = require('./mocks/oneDayCandles');
const monthCandles = require('./mocks/monthCandles');
const yearCandles = require('./mocks/yearCandles');

const date1 = new Date('2018-11-01T00:00:00.000Z'),
  date2 = new Date('2018-12-01T00:00:00.000Z');

const day = interval(CandleInterval.Day1).unsafeGet(),
  minute = interval(CandleInterval.Minute1).unsafeGet(),
  month = interval(CandleInterval.Month1).unsafeGet();

const addMissing1mCandles = addMissingCandles(minute),
  addMissing1dCandles = addMissingCandles(day),
  addMissing1MCandles = addMissingCandles(month);

describe('add missing candles', () => {
  describe('with 1 minute interval', () => {
    it('should add empty candles to list of one candle for 1M period including the 1st day of next M', () => {
      expect(
        pipe(
          groupBy(candle =>
            truncToMinutes(floor(minute, new Date(candle.time_start)))
          ),
          addMissing1mCandles(date1, date2),
          toPairs
        )(oneDayCandles).length
      ).toBe(43201);
    });
    it('should add empty candles to list of thirty candles for 1M period including the 1st day of next M', () => {
      expect(
        pipe(
          groupBy(candle =>
            truncToMinutes(floor(minute, new Date(candle.time_start)))
          ),
          addMissing1mCandles(date1, date2),
          toPairs
        )(monthCandles).length
      ).toBe(43201);
    });
  });

  describe('with 1 day interval', () => {
    it('should add empty candles to list of one candle for 1M period including the 1st day of next M', () => {
      expect(
        pipe(
          groupBy(candle =>
            truncToMinutes(floor(day, new Date(candle.time_start)))
          ),
          addMissing1dCandles(date1, date2),
          toPairs
        )(oneDayCandles).length
      ).toBe(31);
    });

    it('should not add any candles to list of thirty candles for 1M period including the 1st day of next M', () => {
      expect(
        pipe(
          groupBy(candle =>
            truncToMinutes(floor(day, new Date(candle.time_start)))
          ),
          addMissing1dCandles(date1, date2),
          toPairs
        )(monthCandles).length
      ).toBe(31);
    });

    it('should not add candles to list of twenty candles for 1Y period including the last day of month', () => {
      expect(
        pipe(
          groupBy(candle =>
            truncToMinutes(floor(month, new Date(candle.time_start)))
          ),
          addMissing1MCandles(
            new Date('2017-10-01T00:00:00.000Z'),
            new Date('2018-10-01T00:00:00.000Z')
          ),
          toPairs
        )(yearCandles).length
      ).toBe(13);
    });
  });
});

describe('candle monoid', () => {
  it('should calculate 1 candle for 1 interval', () => {
    expect(concatAll(candleMonoid)(monthCandles)).toMatchSnapshot();
  });

  it('should calculate several candles for period with several intervals', () => {
    expect(
      pipe(
        groupBy(candle =>
          truncToMinutes(floor(minute, new Date(candle.time_start)))
        ),
        map(concatAll(candleMonoid))
      )(monthCandles)
    ).toMatchSnapshot();
  });
});

describe('transform candle fn', () => {
  it('take empty candle and returns correctly transformed for result candle', () => {
    expect(
      transformCandle(minute)([date1, candleMonoid.empty])
    ).toMatchSnapshot();
  });

  it('take valid candle and returns correctly transformed for result candle', () => {
    expect(transformCandle(day)([date1, oneDayCandles[0]])).toMatchSnapshot();
  });
});
