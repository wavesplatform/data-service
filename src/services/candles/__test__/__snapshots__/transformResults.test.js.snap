// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`candle monoid should calculate 1 candle for 1 interval 1`] = `
Object {
  "a_dec": 8,
  "close": "0.0001",
  "high": "0.01",
  "low": "1e-8",
  "max_height": 377979,
  "open": "0.01",
  "p_dec": 6,
  "quote_volume": "99.001326",
  "time_start": "2018-11-01T14:03:00.000Z",
  "txs_count": 600,
  "volume": "85260",
  "weighted_average_price": "0.00116116966924700914",
}
`;

exports[`candle monoid should calculate several candles for period with several intervals 1`] = `
Object {
  "2018-11-01T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-01T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-02T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-02T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-03T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-03T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-04T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-04T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-05T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-05T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-06T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-06T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-07T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-07T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-08T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-08T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-09T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-09T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-10T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-10T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-11T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-11T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-12T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-12T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-13T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-13T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-14T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-14T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-15T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-15T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-16T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-16T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-17T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-17T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-18T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-18T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-19T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-19T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-20T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-20T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-21T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-21T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-22T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-22T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-23T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-23T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-24T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-24T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-25T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-25T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-26T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-26T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-27T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-27T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-28T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-28T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-29T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-29T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
  "2018-11-30T14:03:00.000Z": Object {
    "a_dec": 8,
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "max_height": 377979,
    "open": "0.01",
    "p_dec": 6,
    "quote_volume": "3.3000442",
    "time_start": "2018-11-30T14:03:00.000Z",
    "txs_count": 20,
    "volume": "2842",
    "weighted_average_price": "0.00116116966924700914",
  },
}
`;

exports[`transform candle fn take empty candle and returns correctly transformed for result candle 1`] = `
Object {
  "__type": "candle",
  "data": Object {
    "close": null,
    "high": null,
    "low": null,
    "maxHeight": null,
    "open": null,
    "quoteVolume": null,
    "time": 2018-11-01T00:00:00.000Z,
    "timeClose": 2018-11-01T00:00:59.999Z,
    "txsCount": 0,
    "volume": null,
    "weightedAveragePrice": null,
  },
}
`;

exports[`transform candle fn take valid candle and returns correctly transformed for result candle 1`] = `
Object {
  "__type": "candle",
  "data": Object {
    "close": "0.0001",
    "high": "0.01",
    "low": "1e-8",
    "maxHeight": 377979,
    "open": "0.01",
    "quoteVolume": "3.3000442",
    "time": 2018-11-01T00:00:00.000Z,
    "timeClose": 2018-11-01T00:00:59.999Z,
    "txsCount": 20,
    "volume": "2842",
    "weightedAveragePrice": "0.0011",
  },
}
`;
